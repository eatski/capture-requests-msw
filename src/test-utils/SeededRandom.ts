/**
 * シード可能な疑似乱数生成器
 * テスト用途で再現可能なランダム値を生成するために使用
 */
export class SeededRandom {
  private seed: number

  constructor(seed: number) {
    this.seed = seed
  }

  /**
   * Linear Congruential Generator (LCG)
   * 0から1の間の疑似乱数を生成
   */
  next(): number {
    this.seed = (this.seed * 1664525 + 1013904223) % Math.pow(2, 32)
    return this.seed / Math.pow(2, 32)
  }

  /**
   * 指定範囲の整数を生成
   * @param min 最小値（含む）
   * @param max 最大値（含む）
   * @returns min以上max以下の整数
   */
  nextInt(min: number, max: number): number {
    return Math.floor(this.next() * (max - min + 1)) + min
  }

  /**
   * 配列をランダムにシャッフル
   * @param array シャッフルする配列
   * @returns シャッフル済みの新しい配列
   */
  shuffle<T>(array: T[]): T[] {
    const shuffled = [...array]
    for (let i = shuffled.length - 1; i > 0; i--) {
      const j = this.nextInt(0, i)
      ;[shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]]
    }
    return shuffled
  }

}
