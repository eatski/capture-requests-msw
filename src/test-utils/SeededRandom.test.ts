import { describe, it, expect } from 'vitest'
import { SeededRandom } from './SeededRandom'

describe('SeededRandom', () => {

  describe('next()', () => {
    it('should generate deterministic values', () => {
      const rng = new SeededRandom(12345)
      const values = [rng.next(), rng.next(), rng.next()]
      
      expect(values).toMatchSnapshot()
    })

    it('should generate same sequence with same seed', () => {
      const rng1 = new SeededRandom(12345)
      const rng2 = new SeededRandom(12345)
      
      const values1 = [rng1.next(), rng1.next(), rng1.next()]
      const values2 = [rng2.next(), rng2.next(), rng2.next()]
      
      expect(values1).toEqual(values2)
    })

    it('should generate different sequences with different seeds', () => {
      const rng1 = new SeededRandom(12345)
      const rng2 = new SeededRandom(54321)
      
      const values1 = [rng1.next(), rng1.next(), rng1.next()]
      const values2 = [rng2.next(), rng2.next(), rng2.next()]
      
      expect(values1).not.toEqual(values2)
    })

    it('should generate values between 0 and 1', () => {
      const rng = new SeededRandom(12345)
      
      for (let i = 0; i < 100; i++) {
        const value = rng.next()
        expect(value).toBeGreaterThanOrEqual(0)
        expect(value).toBeLessThan(1)
      }
    })
  })

  describe('nextInt()', () => {
    it('should generate deterministic integers', () => {
      const rng = new SeededRandom(12345)
      const values = [
        rng.nextInt(1, 10),
        rng.nextInt(1, 10),
        rng.nextInt(1, 10)
      ]
      
      expect(values).toMatchSnapshot()
    })

    it('should generate integers within specified range', () => {
      const rng = new SeededRandom(12345)
      
      for (let i = 0; i < 100; i++) {
        const value = rng.nextInt(5, 15)
        expect(value).toBeGreaterThanOrEqual(5)
        expect(value).toBeLessThanOrEqual(15)
        expect(Number.isInteger(value)).toBe(true)
      }
    })

    it('should handle single value range', () => {
      const rng = new SeededRandom(12345)
      const value = rng.nextInt(7, 7)
      expect(value).toBe(7)
    })

    it('should generate same sequence with same seed', () => {
      const rng1 = new SeededRandom(12345)
      const rng2 = new SeededRandom(12345)
      
      const values1 = [rng1.nextInt(1, 100), rng1.nextInt(1, 100), rng1.nextInt(1, 100)]
      const values2 = [rng2.nextInt(1, 100), rng2.nextInt(1, 100), rng2.nextInt(1, 100)]
      
      expect(values1).toEqual(values2)
    })
  })

  describe('shuffle()', () => {
    it('should shuffle array deterministically', () => {
      const rng = new SeededRandom(12345)
      const original = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
      const shuffled = rng.shuffle(original)
      
      expect(shuffled).toMatchSnapshot()
    })

    it('should not modify original array', () => {
      const rng = new SeededRandom(12345)
      const original = [1, 2, 3, 4, 5]
      const shuffled = rng.shuffle(original)
      
      expect(original).toEqual([1, 2, 3, 4, 5])
      expect(shuffled).not.toBe(original)
    })

    it('should contain same elements', () => {
      const rng = new SeededRandom(12345)
      const original = [1, 2, 3, 4, 5]
      const shuffled = rng.shuffle(original)
      
      expect(shuffled.sort()).toEqual(original.sort())
    })

    it('should shuffle different arrays with same seed consistently', () => {
      const rng1 = new SeededRandom(12345)
      const rng2 = new SeededRandom(12345)
      
      const shuffled1 = rng1.shuffle([1, 2, 3, 4, 5])
      const shuffled2 = rng2.shuffle([1, 2, 3, 4, 5])
      
      expect(shuffled1).toEqual(shuffled2)
    })

    it('should handle empty array', () => {
      const rng = new SeededRandom(12345)
      const shuffled = rng.shuffle([])
      
      expect(shuffled).toEqual([])
    })

    it('should handle single element array', () => {
      const rng = new SeededRandom(12345)
      const shuffled = rng.shuffle([42])
      
      expect(shuffled).toEqual([42])
    })

    it('should work with different data types', () => {
      const rng = new SeededRandom(12345)
      const strings = ['a', 'b', 'c', 'd']
      const shuffled = rng.shuffle(strings)
      
      expect(shuffled).toMatchSnapshot()
      expect(shuffled.sort()).toEqual(strings.sort())
    })
  })

  describe('randomDelay()', () => {
    it('should return a Promise', () => {
      const rng = new SeededRandom(12345)
      const promise = rng.randomDelay()
      
      expect(promise).toBeInstanceOf(Promise)
    })

    it('should resolve within expected time range', async () => {
      const rng = new SeededRandom(12345)
      const startTime = Date.now()
      
      await rng.randomDelay(10, 20)
      
      const endTime = Date.now()
      const elapsed = endTime - startTime
      
      // Allow some tolerance for execution time
      expect(elapsed).toBeGreaterThanOrEqual(8)
      expect(elapsed).toBeLessThan(30)
    })

    it('should use default values when no parameters provided', async () => {
      const rng = new SeededRandom(12345)
      const startTime = Date.now()
      
      await rng.randomDelay()
      
      const endTime = Date.now()
      const elapsed = endTime - startTime
      
      // Default range is 1-50ms
      expect(elapsed).toBeLessThan(60)
    })

    it('should handle minimum delay', async () => {
      const rng = new SeededRandom(12345)
      const startTime = Date.now()
      
      await rng.randomDelay(0, 0)
      
      const endTime = Date.now()
      const elapsed = endTime - startTime
      
      expect(elapsed).toBeLessThan(10) // Should be very fast
    })
  })

  describe('comprehensive snapshot test', () => {
    it('should generate consistent output across all methods', () => {
      const rng = new SeededRandom(42)
      
      const output = {
        randomNumbers: [rng.next(), rng.next(), rng.next()],
        integers: [rng.nextInt(1, 100), rng.nextInt(1, 100), rng.nextInt(1, 100)],
        shuffledArray: rng.shuffle([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]),
        shuffledStrings: rng.shuffle(['apple', 'banana', 'cherry', 'date', 'elderberry'])
      }
      
      expect(output).toMatchSnapshot()
    })
  })
})