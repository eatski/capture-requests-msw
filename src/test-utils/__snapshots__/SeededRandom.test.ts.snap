// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SeededRandom > comprehensive snapshot test > should generate consistent output across all methods 1`] = `
{
  "integers": [
    23,
    38,
    3,
  ],
  "randomNumbers": [
    0.2523451747838408,
    0.08812504541128874,
    0.5772811982315034,
  ],
  "shuffledArray": [
    1,
    9,
    4,
    10,
    3,
    6,
    8,
    7,
    2,
    5,
  ],
  "shuffledStrings": [
    "date",
    "banana",
    "cherry",
    "elderberry",
    "apple",
  ],
}
`;

exports[`SeededRandom > next() > should generate deterministic values 1`] = `
[
  0.02040268573909998,
  0.01654784823767841,
  0.5431557944975793,
]
`;

exports[`SeededRandom > nextInt() > should generate deterministic integers 1`] = `
[
  1,
  1,
  6,
]
`;

exports[`SeededRandom > shuffle() > should shuffle array deterministically 1`] = `
[
  7,
  3,
  4,
  2,
  9,
  6,
  8,
  5,
  10,
  1,
]
`;

exports[`SeededRandom > shuffle() > should work with different data types 1`] = `
[
  "c",
  "b",
  "d",
  "a",
]
`;
